File: STACMR-matlab/staMRBN.m
function [x, G2, f] = staMRBN (data, E)
% function [x, G2, f] = staMRBN (data, E)
% fits partial order monotonic regression model to binomial data
% data = nsubj x nvar cell array of ncond x 2 matrices of counts (hits, misses) 
% where ncond = no. of conditions
% E is the partial order specified as either:
% (a) cell array, 
% (b) adjacency matrix,
% (c) inequality matrix
% returns:
% x = nsubj x nvar cell array of ncond model means (proportions)
% G2 = nsub-vector of g-squared fit of model
% f = nsub-vector of least squares fit of model
%
% *************************************************************************
% Revised 20 June 2015 following discussions with Oleg Sysoev
% revised 10 March 2017 to incorporate java commands
% *************************************************************************
%
tol = 1e-10;
if nargin < 2; E = []; end
if isempty(E), E = {}; end
if ~iscell(E) && isvector(E), E = {E}; end
if isempty(E)
    disp ('Warning: Partial order undefined.');
end

if ~iscell(data)
    y = BNgen2cell(data); y = binSTATS(y);
elseif isstruct(data{1})
    y = data;
else
    y = binSTATS (data);
end

x = cell(size(y)); f = zeros(size(y)); ml = f; 
for isub = 1:size(y,1)
    for ivar = 1:size(y,2)
        W = y{isub,ivar}.weights; if isvector(W), W = diag(W); end
        [x{isub,ivar}, f(isub,ivar)] = jMR (y{isub,ivar}.means, W, E);
        x{isub,ivar}(x{isub,ivar}<0) = 0;
        x{isub,ivar}(x{isub,ivar}>1) = 1;
        ml(isub,ivar) = mleBN(y{isub,ivar}.count,x{isub,ivar});
    end
end
f = sum(f,2); % sum over variables
G2 = sum(ml,2); G2(G2<=tol)=0;

File: STACMR-R/staMRBN.R
staMRBN = function (data=list(), partial=list()) {
# fits partial order monotonic regression model to binomial data
# data = nsubj list of nvar lists of ncond x 2 matrices of counts (hits, misses) 
# where ncond = no. of conditions
# partial is the partial order specified as either:
# (a) list, 
# (b) adjacency matrix,
# returns:
# x = nsubj x nvar cell array of ncond model means (proportions)
# fval = nsub-vector of least squares fit of model
# gsquare = nsub vector of g-squared fit of model 
#
# *************************************************************************
# converted from Matlab on 20 September 2016
# *************************************************************************
#
  if (missing(partial) | is.null(partial)) {partial = list()}
  if (is.matrix(partial)) {partial = adj2list(partial)} # convert adjacency matrix to list
  
  y = data
  if (is(data,"data.frame")) {y = gen2listBN(data)} # convert data frame to list of lists
  a = "means" %in% variable.names(as.data.frame(y))
  if (!a) {y = binSTATS(y)}

  #y = binSTATS(data) # get summary stats
  
  x = vector('list', y$ngroup); f = matrix(0, y$ngroup, y$nvar); g = f
  
  for (igroup in 1:y$ngroup) {
    temp = matrix(0, y$ncond, y$nvar)
    for (ivar in 1:y$nvar) {
      means = y[[igroup]][[ivar]]$means
      weights = y[[igroup]][[ivar]]$weights
      count = y[[igroup]][[ivar]]$count
      
      out = jMR (means, weights, partial) # do monotonic regression
      
      out$x[which(out$x < 0)] = 0 # lower limit of 0
      out$x[which(out$x > 1)] = 1 # upper limit of 1
      
      temp[,ivar] = out$x
      f[igroup,ivar] = out$fval
      g[igroup,ivar] = mleBN (out$x, count)
    }
    x[[igroup]] = temp
  }
  f = as.matrix(rowSums(f))
  g = as.matrix(rowSums(g))
  return(list(x=x, fval=f, g.squared=g))
}

mleBN = function (x, count) {
  # calculates g.squared statistic
  nn = sum(count)
  x[which(x <= 0)] = .5/nn
  x[which(x >= 1)] = (nn - .5)/nn
  n = rowSums (count)
  a = t(rbind(x, 1-x))*cbind(n, n)
  g2 = 2*count*log(count/a); g2[which(count <= 0)]=0;
  f = sum(g2); f[which(f < 0)] = 0
  return(f)
}
